core/chains/evm/gas/universal_estimator.go

```diff
@@ -0,0 +1,412 @@
+	}
+	if fee.FeeCap.Cmp(maxPrice) > 0 {
+		u.logger.Warnf("estimated maxFeePerGas: %v is greater than the maximum price configured: %v, returning the maximum price instead.",
+			fee.FeeCap, maxPrice)
+		fee.FeeCap = maxPrice
+		if fee.TipCap.Cmp(maxPrice) > 0 {
+			u.logger.Warnf("estimated maxPriorityFeePerGas: %v is greater than the maximum price configured: %v, returning the maximum price instead.",
+				fee.TipCap, maxPrice)
+			fee.TipCap = maxPrice
+		}
+	}
+
+	return
+}
+
+// FetchDynamicPrice uses eth_feeHistory to fetch the basFee of the latest block and the Nth maxPriorityFeePerGas percentiles
+// of the past X blocks. It also fetches the highest Zth maxPriorityFeePerGas percentile of the past X blocks. Z is configurable
+// and it represents the highest percentile we're willing to pay.
+// A buffer is added on top of the latest basFee to catch fluctuations in the next blocks. On Ethereum the increase is baseFee*1.125 per block.
+func (u *UniversalEstimator) FetchDynamicPrice() (fee DynamicFee, err error) {
+	ctx, cancel := u.stopCh.CtxCancel(evmclient.ContextWithDefaultTimeout())
+	defer cancel()
+
+	if u.config.BlockHistorySize == 0 {
+		return fee, fmt.Errorf("BlockHistorySize cannot be 0")
+	}
+	// RewardPercentile will be used for maxPriorityFeePerGas estimations and connectivityPercentile to set the highest threshold for bumping.
+	feeHistory, err := u.client.FeeHistory(ctx, u.config.BlockHistorySize, []float64{u.config.RewardPercentile, ConnectivityPercentile})
+	if err != nil {
+		return fee, fmt.Errorf("failed to fetch dynamic prices: %s", err)
+	}
+
+	// Latest base fee
+	baseFee := (*assets.Wei)(feeHistory.BaseFee[len(feeHistory.BaseFee)-1])
+	latestBlock := big.NewInt(0).Add(feeHistory.OldestBlock, big.NewInt(int64(u.config.BlockHistorySize)))
+
+	priorityFee := big.NewInt(0)
+	priorityFeeThreshold := big.NewInt(0)
+	for _, fee := range feeHistory.Reward {
+		priorityFee = priorityFee.Add(priorityFee, fee[0])
+		// We don't need an average, we need the max value
+		priorityFeeThreshold = bigmath.Max(priorityFeeThreshold, fee[1])
+	}
+
+	u.priorityFeeThresholdMu.Lock()
+	u.priorityFeeThreshold = (*assets.Wei)(priorityFeeThreshold)
+	u.priorityFeeThresholdMu.Unlock()
+
+	maxPriorityFeePerGas := (*assets.Wei)(priorityFee.Div(priorityFee, big.NewInt(int64(u.config.BlockHistorySize))))
+	// baseFeeBufferPercentage is used as a safety to catch fluctuations in the next block.
+	maxFeePerGas := baseFee.AddPercentage(BaseFeeBufferPercentage).Add((maxPriorityFeePerGas))
+
+	promUniversalEstimatorBaseFee.WithLabelValues(u.chainID.String()).Set(float64(baseFee.Int64()))
+	promUniversalEstimatorMaxPriorityFeePerGas.WithLabelValues(u.chainID.String()).Set(float64(maxPriorityFeePerGas.Int64()))
+	promUniversalEstimatorMaxFeePerGas.WithLabelValues(u.chainID.String()).Set(float64(maxFeePerGas.Int64()))
+
+	u.logger.Debugf("Fetched new dynamic prices, block#: %v - maxFeePerGas: %v - maxPriorityFeePerGas: %v - maxPriorityFeeThreshold: %v",
```

Comment:
Feel like it may be worthwhile to also include the oldest block here. I know it can be inferred from the block history size but it may be easier to get a snapshot of the block window in one log line this way